File: lib/ts/src/internal/types/VNumber.ts
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

import { ValidationFailure } from '../validation/ValidationFailure';
import { VType } from './VType';
import { validateNumber } from '../validation/ValidateNumber';
import { generateRandomNumber } from '../generation/GenerateRandomNumber';
import { GenerateContext } from '../generation/GenerateContext';
import { VTypeDeclaration } from './VTypeDeclaration';
import { ValidateContext } from '../validation/ValidateContext';

export const numberName: string = 'Number';

export class VNumber extends VType {
    getTypeParameterCount(): number {
        return 0;
    }

    validate(value: any, typeParameters: VTypeDeclaration[], ctx: ValidateContext): ValidationFailure[] {
        return validateNumber(value);
    }

    generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: VTypeDeclaration[],
        ctx: GenerateContext,
    ): any {
        return generateRandomNumber(blueprintValue, useBlueprintValue, ctx);
    }

    getName(): string {
        return numberName;
    }
}

File: lib/py/msgpact/internal/types/VNumber.py
from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from ..validation.ValidateContext import ValidateContext
    from ...RandomGenerator import RandomGenerator
    from .VTypeDeclaration import VTypeDeclaration
    from ..validation.ValidationFailure import ValidationFailure
    from ..generation.GenerateContext import GenerateContext


from .VType import VType

_NUMBER_NAME: str = "Number"


class VNumber(VType):

    def get_type_parameter_count(self) -> int:
        return 0

    def validate(self, value: object,
                 type_parameters: list['VTypeDeclaration'], ctx: 'ValidateContext') -> list['ValidationFailure']:
        from ..validation.ValidateNumber import validate_number
        return validate_number(value)

    def generate_random_value(self, blueprint_value: object, use_blueprint_value: bool, type_parameters: list['VTypeDeclaration'], ctx: 'GenerateContext') -> object:
        from ..generation.GenerateRandomNumber import generate_random_number
        return generate_random_number(blueprint_value, use_blueprint_value, ctx)

    def get_name(self) -> str:
        return _NUMBER_NAME
